posts/2024-10-17-Rust-1.82.0.md

```diff
@@ -0,0 +1,319 @@
+
+For several years, the macros `std::ptr::addr_of!` and `std::ptr::addr_of_mut!` have served this purpose. Now the time has come to provide a proper native syntax for this operation: `addr_of!(expr)` becomes `&raw const expr`, and `addr_of_mut!(expr)` becomes `&raw mut expr`. For example:
+
+```rust
+#[repr(packed)]
+struct Packed {
+    not_aligned_field: i32,
+}
+
+fn main() {
+    let p = Packed { not_aligned_field: 1_82 };
+
+    // This would be undefined behavior!
+    // It is rejected by the compiler.
+    //let ptr = &p.not_aligned_field as *const i32;
+
+    // This is the old way of creating a pointer.
+    let ptr = std::ptr::addr_of!(p.not_aligned_field);
+
+    // This is the new way.
+    let ptr = &raw const p.not_aligned_field;
+
+    // Accessing the pointer has not changed.
+    // Note that `val = *ptr` would be undefined behavior because
+    // the pointer is not aligned!
+    let val = unsafe { ptr.read_unaligned() };
+}
+```
+
+The native syntax makes it more clear that the operand expression of these operators is interpreted as a [place expression](https://www.ralfj.de/blog/2024/08/14/places.html). It also avoids the term "address-of" when referring to the action of creating a pointer. A pointer is [more than just an address](https://rust-lang.github.io/rfcs/3559-rust-has-provenance.html), so Rust is moving away from terms like "address-of" that reaffirm a false equivalence of pointers and addresses.
+
+### Safely addressing unsafe `static`s
+
+This code is now allowed:
+
+```rust
+static mut STATIC_MUT: Type = Type::new();
+extern "C" {
+    static EXTERN_STATIC: Type;
+}
+fn main() {
+     let static_mut_ptr = &raw mut STATIC_MUT;
+     let extern_static_ptr = &raw const EXTERN_STATIC;
+}
+```
+
+In an expression context, `STATIC_MUT` and `EXTERN_STATIC` are [place expressions](https://doc.rust-lang.org/reference/expressions.html#place-expressions-and-value-expressions). Previously, the compiler's safety checks were not aware that the raw ref operator did not actually affect the operand's place, treating it as a possible read or write to a pointer. No unsafety is actually present, however, as it just creates a pointer.
+
+Relaxing this may cause problems where some unsafe blocks are now reported as unused if you deny the `unused_unsafe` lint, but they are now only useful on older versions. Annotate these unsafe blocks with `#[allow(unused_unsafe)]` if you wish to support multiple versions of Rust, as in this example diff:
+
+```diff
+ static mut STATIC_MUT: Type = Type::new();
+ fn main() {
++    #[allow(unused_unsafe)]
+     let static_mut_ptr = unsafe { std::ptr::addr_of_mut!(STATIC_MUT) };
+ }
+```
+
+A future version of Rust is expected to generalize this to other expressions which would be safe in this position, not just statics.
+
```

Comment:
````suggestion

### Unsafe extern

Rust code can use functions and statics from foreign code.  The type signatures of these foreign items are provided in `extern` blocks.  Historically, all items within `extern` blocks have been unsafe to call, but we didn't have to write `unsafe` anywhere on the `extern` block itself.

However, if a signature within the `extern` block is incorrect, then using that item will result in undefined behavior.  Would that be the fault of the person who wrote the `extern` block, or the person who used that item?

We've decided that it's the responsibility of the person writing the `extern` block to ensure that all signatures contained within it are correct, and so we now allow writing `unsafe extern`:

```rust
unsafe extern {
    pub safe fn sqrt(x: f64) -> f64;
    pub unsafe fn strlen(p: *const u8) -> usize;
}
```

One benefit of this is that items within an `unsafe extern` block can be marked as safe to call.  In the above example, we can call `sqrt` without using `unsafe`.  Items that aren't marked with either `safe` or `unsafe` are conservatively assumed to be `unsafe`.

In future releases, we'll be encouraging the use of `unsafe extern` with lints.  Starting in Rust 2024, using `unsafe extern` will be required.

For further details, see [RFC 3484](https://github.com/rust-lang/rfcs/blob/master/text/3484-unsafe-extern-blocks.md) and the ["Unsafe extern blocks"](https://doc.rust-lang.org/nightly/edition-guide/rust-2024/unsafe-extern.html) chapter of the edition guide.

````

Reply:
Unsafe extern is another major user-visible addition in Rust 1.82.  I'd put this ahead of unsafe attributes, and move it further up along with unsafe attributes.

Reply:
Maybe we can raise the `safe` part in the section title? That's a big benefit IMO, and I missed it at first glance. How about "Unsafe extern and safe calls"?

Reply:
Good point.  Maybe "Unsafe extern and safe items".

Reply:
Or maybe more pithily, "Safe items in unsafe extern".